### PR TITLE
Feature/ Challenge based on rounds

### LINKED ISSUE #2399: process challenges based on round
- validate that create-challenge transaction is included in the block with exact round
- process challenges based on round created and round response submitted, make them expire based on current round as well

### PR DESCRIPTION
**Pull request by Jayashsatolia403**
### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- 0chain:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 1]
```diff
@@ -191,12 +193,30 @@ func (c *ChallengeEntity) getCommitTransaction() (*transaction.Transaction, erro
 	tx := datastore.GetStore().GetTransaction(ctx)
 
 	createdTime := common.ToTime(c.CreatedAt)
+
+	logging.Logger.Info("[challenge]verify: ",
+		zap.Any("challenge_id", c.ChallengeID),
+		zap.Time("created", createdTime))
+
+	network := zcncore.GetNetwork()
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/challenge/challenge.go]: move this call into the gosdk

File: code/go/0chain.net/blobbercore/stats/blobberstats.go [hunk 2]
```diff
@@ -306,21 +306,14 @@ func (bs *BlobberStats) loadAllocationStats(ctx context.Context) {
...
 	}
 
-	if err = rows.Err(); err != nil && err != sql.ErrNoRows {
+	if err = rows.Err(); err != nil && !errors.Is(err, sql.ErrNoRows) {
 		Logger.Error("Error in scanning record for blobber stats",
 			zap.Error(err))
 		return
 	}
 
-	var count int64
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/stats/blobberstats.go:315]: not sure I understand this change, can you elaborate please

Comment (Jayashsatolia403) [reply code/go/0chain.net/blobbercore/stats/blobberstats.go:315]: I changed this to fix the empty pages issue in diagnostic pages.

File: code/go/0chain.net/blobbercore/challenge/challenge.go (hunk 1 above)
Comment (Jayashsatolia403) [reply code/go/0chain.net/blobbercore/challenge/challenge.go]: We need to getNetwork here only. I tried moving in gosdk and we are getting nil.

File: code/go/0chain.net/core/transaction/entity.go [hunk 3]
```diff
@@ -139,6 +137,22 @@ func (t *Transaction) GetTransaction() zcncore.TransactionScheme {
 
 func (t *Transaction) ExecuteSmartContract(address, methodName string, input interface{}, val uint64) error {
 	t.wg.Add(1)
+
+	logging.Logger.Info("Jayash", zap.Any("address", address), zap.Any("methodName", methodName), zap.Any("input", input), zap.Any("val", val))
```
Comment (dabasov) [inline code/go/0chain.net/core/transaction/entity.go]: clean up this log

File: code/go/0chain.net/core/transaction/entity.go [hunk 4]
```diff
@@ -139,6 +137,22 @@ func (t *Transaction) GetTransaction() zcncore.TransactionScheme {
...
+
+	sn := transaction.SmartContractTxnData{Name: methodName, InputArgs: input}
+	snBytes, err := json.Marshal(sn)
+	if err != nil {
+		logging.Logger.Error("Jayash", zap.Error(err))
```
Comment (dabasov) [inline code/go/0chain.net/core/transaction/entity.go]: and this

File: code/go/0chain.net/validatorcore/storage/challenge_handler_main.go [hunk 5]
```diff
@@ -9,5 +9,12 @@ import (
 )
 
 func ChallengeHandler(ctx context.Context, r *http.Request) (interface{}, error) {
-	return challengeHandler(ctx, r)
+	res, err := challengeHandler(ctx, r)
+
+	if len(Last5Transactions) >= 5 {
```
Comment (dabasov) [inline code/go/0chain.net/validatorcore/storage/challenge_handler_main.go:14]: we should track all the transactions we send to 0chain, shouldn't we?

File: code/go/0chain.net/core/transaction/entity.go [hunk 6]
```diff
@@ -16,6 +16,8 @@ import (
 	"github.com/0chain/blobber/code/go/0chain.net/core/node"
 )
 
+var Last50Transactions []string
```
Comment (dabasov) [inline code/go/0chain.net/core/transaction/entity.go]: it should be guarded with locks

**Review by dabasov — CHANGES_REQUESTED** [review]

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**